boxsdk/object/upload_session.py

```diff
@@ -0,0 +1,146 @@
+# coding: utf-8
+from __future__ import unicode_literals, absolute_import
+
+import base64
+import hashlib
+import json
+
+from .base_object import BaseObject
+from ..config import API
+
+
+class UploadSession(BaseObject):
+    _item_type = 'upload_session'
+    _parent_item_type = 'file'
+
+    def get_url(self, *args):
+        """
+        Base class override. Endpoint is a little different - it's /files/upload_sessions.
+
+        :rtype:
+            `unicode`
+        """
+        return self.session.get_url(
+            '{0}s/{1}s'.format(self._parent_item_type, self._item_type),
+            self._object_id,
+            *args
+        ).replace(API.BASE_API_URL, API.UPLOAD_URL)
+
+    def get_parts(self):
+        """
+        Get a list of parts uploaded so far.
+
+        :returns:
+            Returns a `list` of parts uploaded so far.
+        :rtype:
+            `list` of `dict`
+        """
+        response = self.session.get(self.get_url('parts')).json()
+        return response['entries']
+
+    def _calculate_part_sha1(self, content_stream):
+        """
+        Calculate the SHA1 hash of the chunk stream for a given part of the file.
+
+        :returns:
+            The unencoded SHA1 hash of the part.
+        :rtype:
+            `bytes`
+        """
+        content_sha1 = hashlib.sha1()
+        stream_position = content_stream.tell()
+        hashed_length = 0
+        while hashed_length < self.part_size:  # pylint:disable=no-member
+            chunk = content_stream.read(self.part_size - hashed_length)  # pylint:disable=no-member
+            if chunk is None:
+                continue
+            if not chunk:
+                break
+            hashed_length += len(chunk)
+            content_sha1.update(chunk)
+        content_stream.seek(stream_position)
+        return content_sha1.digest()
+
+    def upload_part(self, content_stream, offset, total_size, part_content_sha1=None):
+        """
+        Upload a part of a file.
+
+        :param content_stream:
+            File-like object containing the content of the part to be uploaded.
+        :type content_stream:
+            :class:`File`
+        :param offset:
+            Offset, in number of bytes, of the part compared to the beginning of the file.
+        :type offset:
+            `int`
+        :param total_size:
+            The size of the file that this part belongs to.
+        :type total_size:
+            `int`
+        :param part_content_sha1:
+            SHA-1 hash of the part's content. If not specified, this will be calculated.
+        :type part_content_sha1:
+            `unicode`
+        :returns:
+            The uploaded part.
+        :rtype:
+            `dict`
+        """
+        if part_content_sha1 is None:
+            part_content_sha1 = self._calculate_part_sha1(content_stream)
+
+        range_end = min(offset + self.part_size - 1, total_size - 1)  # pylint:disable=no-member
+
+        return self._session.put(
```

Comment:
It might be good to return just the actual part `dict` here, to make it easier for developers to form the correct array for `commit()`

Reply:
That would be a nice method on a `ChunkedUploadPart` object.